text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,245 @@
+
+Now rustup could be augmented to build stdlib binaries in addition to downloading them, but we have extra configuration options for the standard library such as [panic strategies](https://github.com/rust-lang/rfcs/blob/master/text/1513-less-unwinding.md) along with plans to add more such as [`core` without floats](https://github.com/rust-lang/rfcs/issues/1364).
+One could also cobble together a way to to tell rustup such configuration options, but we'd like to make sure that all dependencies agree with the plan before trying to execute it.
+For panicking, packages can already set a `profile.dev.panic` option to require a specific strategy, and for float support we should add a (default opt-in) feature to core.
+Then, if packages can explicitly depend on core they can also specify whether the float feature is needed.
+Cargo would be able to infer both of these options by inspecting all crates in the dependency graph.
+
+Rustbuild must currently perform multiple `cargo builds`, the first to build the standard library and the rest to build things which depend on the standard library.
+If rustc, rustfmt, etc, and their deps (some of which come from crates.io, and thus aren't specially tailored for building rust) declare deps on std, rustbuild wouldn't need multiple lockfiles.
+Keeping multiple lockfiles in sync is nuisance, and this gets us one step closer to a single `cargo build` building rustc.
+
+The use-cases so far mainly benefit niche corners of the Rust community, but the last should be useful for just about everyone.
+Now that multiple versions of Rust have been released, it can be useful to specify the minimum version.
+If and when Rust 2, a version with breaking changes, comes out, this will be all the more important.
+We don't yet have a plan yet to track the language itself, but by tracking standard library dependencies we make it trivial to specify version requirements like any other package.
+
+
+# Detailed design
+
+## Standard library dependencies
+
+First and foremost, one will now be able to explicitly on standard library, e.g. with `std = { version = "1.10", stdlib = true }`.
+From the users's perspective, `stdlib = true` simply indicates that the dependendend on crate is from the standard library.
+The version for stdlib crates comes from the version of Rust their interfaces are defined in.
+A version requirement must be specified.
+The full beadth of options available with our existing dependencies, e.g. features and overrides, will be supported.
+
+## Implicit dependencies
+
+For backwards compatibility, Cargo must inject such standard library dependencies for existing packages.
+These injected standard library dependencies are called "implicit dependencies" because the user does not specify them explicitly.
+We have an obligation to not break packages depending only on stable interfaces, so the implicit dependencies will include both `std` and `core`:
```

Comment:
Small user experience concern:

The current behavior of Cargo is to compile all the crates listed under `dependencies`, even if they are not used (`extern crate`d) by the crate that's being build. That behavior paired with this implicit dependencies scheme can lead to surprising behavior (extra compilation time or even compiler errors) because of false expectations:

```
$ cargo new foo
# Add #![no_std]
$ edit src/lib.rs
$ cargo build --target custom
# ^ One would expect that this only compiles `core` but instead it compiles *all* the crates behind `std`
```

As per this RFC, the solution is to opt-out of the implict `std` dependency but this is not obvious (unless you've read this RFC :-). I expect that the next thing the user will try is to explicitly list the `core` under dependencies but this won't solve the problem either.

I can't think of a way to avoid this situation by just modifying the implicit dependencies scheme, but a possible (more general) solution would be to make Cargo smarter and have it check if a dependency is actually used before compiling it.